src/fastify-controller.js

```diff
@@ -0,0 +1,91 @@
+import { promisify } from 'util'
+import StateMachine from 'start-stop-state-machine'
+
+/**
+ * @typedef {Object} StartOpts
+ *
+ * @property {string} [host]
+ * @property {number} [port]
+ */
+
+// Class to properly manage the server lifecycle of a Fastify instance
+export class FastifyController {
+  #fastify
+  #fastifyStarted
+  #host
+  #port
+  #serverState
+
+  /**
+   * @param {Object} opts
+   * @param {import('fastify').FastifyInstance} opts.fastify
+   */
+  constructor({ fastify }) {
+    this.#fastifyStarted = false
+    this.#host = '127.0.0.1'
+    this.#port = 0
+
+    this.#fastify = fastify
+
+    this.#serverState = new StateMachine({
+      start: this.#startServer.bind(this),
+      stop: this.#stopServer.bind(this),
+    })
+  }
+
+  /**
+   * @param {StartOpts} [opts]
+   */
+  async #startServer({ host = '127.0.0.1', port = 0 } = {}) {
+    this.#host = host
+    this.#port = port
```

Comment:
Nit: Again, something carried over, but I don't think these need to be instance properties - they are only used within this method.

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/440/commits/f80ac9a6490da9e3a2f23510d911e88f369ebf5a